core/chains/solana/chain.go

```diff
@@ -44,13 +46,133 @@ type chain struct {
 	lggr           logger.Logger
 
 	// tracking node chain id for verification
-	clientCache map[string]cachedClient // map URL -> {client, chainId} [mainnet/testnet/devnet/localnet]
+	clientCache map[string]*verifiedCachedClient // map URL -> {client, chainId} [mainnet/testnet/devnet/localnet]
 	clientLock  sync.RWMutex
 }
 
-type cachedClient struct {
-	id string
-	rw solanaclient.ReaderWriter
+type verifiedCachedClient struct {
+	chainID         string
+	expectedChainID string
+	nodeURL         string
+
+	lggr logger.Logger
+
+	chainIDVerified     bool
+	chainIDVerifiedLock sync.RWMutex
+
+	solanaclient.ReaderWriter
+}
+
+func (v *verifiedCachedClient) verifyChainID() (bool, error) {
+	v.chainIDVerifiedLock.RLock()
+	if v.chainIDVerified {
+		defer v.chainIDVerifiedLock.RUnlock()
+		return true, nil
+	}
+	v.chainIDVerifiedLock.RUnlock()
+
+	var err error
+
+	v.chainIDVerifiedLock.Lock()
+	defer v.chainIDVerifiedLock.Unlock()
+
+	v.chainID, err = v.ReaderWriter.ChainID()
+	if err != nil {
+		v.chainIDVerified = false
+		return v.chainIDVerified, errors.Wrap(err, "failed to fetch ChainID in verifiedCachedClient")
+	}
+
+	// check chainID matches expected chainID
+	expectedChainID := strings.ToLower(v.expectedChainID)
+	if v.chainID != expectedChainID {
+		v.chainIDVerified = false
+		return v.chainIDVerified, errors.Errorf("client returned mismatched chain id (expected: %s, got: %s): %s", expectedChainID, v.chainID, v.nodeURL)
+	}
+
+	v.chainIDVerified = true
+
+	return v.chainIDVerified, nil
+}
+
+func (v *verifiedCachedClient) SendTx(ctx context.Context, tx *solanago.Transaction) (solanago.Signature, error) {
+	verified, err := v.verifyChainID()
```

Comment:
Would it not make more sense to simply verify the chain once, on `Start`?

Reply:
That was the actual issue, that verifying the chain on start was preventing the struct from being created at all. So what we are doing now is to allow the struct to be created and allow for the RPC methods to work if eventually the chain id can be verified.